Feature [Write API]: Insert Class Section Events Operation

jmchamica opened the pull request:
Implementation decisions:
- 3 Stored Procedures for creation/replacement of rows
  - declaring 3 SPs instead of one big SP saves extra insert/update/delete operations by applying changes to the more broad parameters (e.g. the Programme table is updated at the beginning of the transaction and not every time an Event is inserted)
- Jackson delivers a `JsonNode` object to the `Controller` for the purpose of JSON Schema validation of the HTTP request
- The JSON Schema is very strict on the format of the requests. The idea is to prevent unpredictable or undetectable errors when calling the write API operation

Closes #144 #128

jmchamica committed:
[Write API] Adds Procedure for insertion of entities.

jmchamica committed:
[Write API]: Completes procedure for Course,Class,ClassSection insertion.

jmchamica committed:
[Write API]: Replaces some procedure for functions.

Functions allow conditional call without using IF blocks.

jmchamica committed:
[Write API]: Delete all events from the same category (class section)

Related to the insertClassSectionEvents Write API operation.

jmchamica committed:
[Write API]: Insertion of class section events (Stored Procedure)

jmchamica committed:
Merge branch 'bug/gh-145-fixInsertSqlEventCategory' into feature/gh-128-insertClassSectionEvents-db

jmchamica committed:
Fixes build errors

jmchamica committed:
Merge branch 'master' into feature/gh-144-insertClassSectionEvents-api

jmchamica committed:
Categories are statically defined. Make no attempt to create new ones.

jmchamica committed:
InsertClassSectionEvents controller with mocked values.

jmchamica committed:
Merge branch 'feature/gh-128-insertClassSectionEvents-db' into feature/gh-144-insertClassSectionEvents-api

jmchamica committed:
[Write API]: Complete Data classes (InsertClassSectionEvents operation)

jmchamica committed:
Interpret and validate HTTP request body.

jmchamica committed:
Correcting WriteAPI Controller's errors.

jmchamica committed:
Merge branch 'master' into feature/gh-144-insertClassSectionEvents-api

jmchamica committed:
Merging with master.

jmchamica committed:
[Write API]: Adjustments

jmchamica committed:
Merge branch 'master' into feature/gh-144-insertClassSectionEvents-api

jmchamica committed:
[Write API]: Include event category and language in the JSON request.

Category is mandatory while Language is optional (the default en-US will
be used instead).

JorgePBrown commented on project/src/main/kotlin/org/ionproject/core/kotlinx.kt:
```diff
@@ -78,3 +78,7 @@ fun PGobject.split(): List<String> {
     return list
 }
+
+fun List<String>.join(separator: String) = this.reduceRight { l, r -> "${l}${separator}${r}" }
+fun List<String>.join() = this.join(",")
```
What is the point of this? Isn't it the same as [Iterable::joinToString](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/join-to-string.html)?

jmchamica replied:
You're right, I couldn't remember the name of the function so I just made one real quick. I'll fix it.

JorgePBrown commented on project/src/main/kotlin/org/ionproject/core/writeApi/insertClassSectionEvents/InsertClassSectionEventsRepoImpl.kt:
```diff
@@ -0,0 +1,106 @@
+      .invoke()
+  }
+
+  override fun transaction(t: (InsertClassSectionEventsRepo) -> Unit): Boolean {
+    tm.run { t(this) }
+    return true
+  }
```
What is this for exactly? Why should someone do
```kotlin
    try {
      repo.transaction { sql ->
        sql.insertClassSectionSchoolInfo(
          schoolInfo.schoolName,
          schoolInfo.schoolAcr,
          schoolInfo.programmeName,
          schoolInfo.programmeAcr,
          schoolInfo.programmeTermSize,
          schoolInfo.calendarTerm
        )

        schoolInfo.courses.forEach { course ->
          sql.insertClassSectionCourseInfo(
            course.name,
            course.acr,
            schoolInfo.calendarSection,
            schoolInfo.calendarTerm,
            schoolInfo.language,
            schoolInfo.category
          )
      }
    } catch() {
      ...
    }
```

instead of: 

```kotlin
try {
      repo.insertClassSectionSchoolInfo(
        schoolInfo.schoolName,
        schoolInfo.schoolAcr,
        schoolInfo.programmeName,
        schoolInfo.programmeAcr,
        schoolInfo.programmeTermSize,
        schoolInfo.calendarTerm
      )

      schoolInfo.courses.forEach { course ->
        repo.insertClassSectionCourseInfo(
        course.name,
        course.acr,
        schoolInfo.calendarSection,
        schoolInfo.calendarTerm,
        schoolInfo.language,
        schoolInfo.category
      )
    } catch() {
      ...
    }
```

Maybe it's something to do with closures that I'm not seeing, just need help to understand.

jmchamica replied:
The function encapsulates all calls in a common transaction. This way, each and every change of the whole operation is guaranteed to rollback in case one of the Procedures fail.

JorgePBrown replied:
I'm stupid. Ignore this.

JorgePBrown commented on project/src/test/resources/docker/sql/create-schema.sql:
```diff
@@ -281,24 +281,24 @@ CREATE OR REPLACE PROCEDURE dbo.newEvent(
-    byday VARCHAR(20),
-    until TIMESTAMP,
-    stamp_time TIMESTAMP DEFAULT now()
+  dtend TIMESTAMP,
+  dtstart_dtend_type INT,
+  location VARCHAR(128),
+  byday VARCHAR(20),
+  until TIMESTAMP,
+  stamp_time TIMESTAMP DEFAULT now()
```
We really need a style checker.

jmchamica replied:
Indeed. Although I'm not sure style checkers would examine SQL scripts.

JorgePBrown commented on project/src/test/resources/docker/sql/create-schema.sql:
```diff
@@ -469,3 +475,223 @@ CREATE TABLE dbo.policies(
+
+  -- Insert or update Programme
+  -- coalesce: if any of the supplied parameters is null, use the previous value of the column instead
+  UPDATE
+    dbo.programme
+  SET
+    name     = COALESCE(programmeName, name),
+    acronym  = COALESCE(programmeAcr, acronym),
+    termSize = COALESCE(programmeTermSize, termSize)
+  WHERE
+    name = programmeName OR acronym = programmeAcr;
+
+  INSERT INTO
+    dbo.Programme(name, acronym, termSize)
+  SELECT
+    programmeName, programmeAcr, programmeTermSize
+  WHERE
+    NOT EXISTS(
+      SELECT id FROM dbo.Programme p
+      WHERE p.name = programmeName OR p.acronym=programmeAcr
+    );
```
I think this could be made more clear with an `INSERT ON CONFLICT` clause.
```sql
INSERT INTO
  dbo.programme(name, acronym, termSize)
VALUES
 (
    COALESCE(programmeName, name),
    COALESCE(programmeAcr, acronym),
    COALESCE(programmeTermSize, termSize)
 )
ON CONFLICT DO UPDATE 
  SET
    name     = COALESCE(programmeName, name),
    acronym  = COALESCE(programmeAcr, acronym),
    termSize = COALESCE(programmeTermSize, termSize);
``` 

I'm not sure the statement would be exactly like this but it's something close.

JorgePBrown commented on project/src/test/resources/docker/sql/create-schema.sql:
```diff
@@ -469,3 +475,223 @@ CREATE TABLE dbo.policies(
+  INSERT INTO
+    dbo.CalendarTerm(id, start_date, end_date)
+  SELECT
+    calendarTerm, NULL, NULL
```
Are the `start_date` and the `end_date` never supplied by the Integration project? Do they always need to be added manually?

jmchamica replied:
Good catch. I had postponed this decision until the implementation was close to being complete. Thanks for reminding me.

jmchamica replied:
In the previous meeting, a decision was made to implement a dependency between and recurrent event's starting and end dates with the Calendar Term.

JorgePBrown commented on project/src/test/resources/docker/sql/create-schema.sql:
```diff
@@ -469,3 +475,223 @@ CREATE TABLE dbo.policies(
+    RAISE 'For the Course parameters, you must provide at least a "name" or an "acronym"';
+  END IF;
+
+  SELECT id INTO langid FROM dbo.Language WHERE name = lang;
+  IF (langid IS NULL) THEN
+    RAISE '% language does not exist', lang;
+  END IF;
+
+  SELECT
+    c.id INTO categoryid
+  FROM
+    dbo.category c JOIN
+    dbo.categorylanguage cl ON c.id = cl.category
+  WHERE
+    name = categ AND language = langid;
+  IF (categoryid IS NULL) THEN
+    RAISE '"%" is not a valid category for the language %', categ, lang;
+  END IF;
```
Is this for checking if a category with the course name exists or what kind of category should this be?

jmchamica replied:
This is checking if the category of the event specified in the HTTP request is valid. Since categories are directly inserted to the database and never modified by the write API.

JorgePBrown replied:
Does the Integration project choose the categories of the event then? Do they know which ones exist currently?

jmchamica replied:
We decided to use the Category ID (Integer) and list the available IDs in the Write API documentation.

JorgePBrown commented on project/src/test/resources/docker/sql/create-schema.sql:
```diff
@@ -469,3 +475,223 @@ CREATE TABLE dbo.policies(
+
+  -- Insert or update Programme
+  -- coalesce: if any of the supplied parameters is null, use the previous value of the column instead
+  UPDATE
+    dbo.course
+  SET
+    name     = COALESCE(courseName, name),
+    acronym  = COALESCE(courseAcr, acronym)
+  WHERE
+    name = courseName OR acronym = courseAcr;
+
+  INSERT INTO
+    dbo.Course(name, acronym)
+  SELECT
+    courseName, courseAcr
+  WHERE
+    NOT EXISTS(
+      SELECT id FROM dbo.course
+      WHERE name = courseName OR acronym = courseAcr
+    );
```
Same thing with `INSERT ON CONFLICT` clause.

JorgePBrown commented on project/src/test/resources/docker/sql/create-schema.sql:
```diff
@@ -469,3 +475,223 @@ CREATE TABLE dbo.policies(
+    
+  SELECT id INTO clid FROM dbo.Class WHERE courseId = cid AND calendarTerm = calTerm;
+
+  -- Creating the Class
```
```suggestion
  -- Creating the ClassSection
```

JorgePBrown commented on project/src/test/resources/docker/sql/create-schema.sql:
```diff
@@ -469,3 +475,223 @@ CREATE TABLE dbo.policies(
+	summary VARCHAR(50),
+	description VARCHAR(200),
+  lang VARCHAR(10),
+	categ VARCHAR(100),
```
An event can have multiple categories. Shouldn't this be a `VARCHAR(100)[]`?

jmchamica replied:
The `insertClassSectionEvents` operation limits this option by restricting it to only one category. The suggested operation by the Integration Group was actually a `insertClassSectionSchedule` operation which would always create `Lecture` events.

TLDR: we could apply this change, but for now I'd like to keep it simple and maybe patch it later.

JorgePBrown commented on project/src/test/resources/docker/sql/create-schema.sql:
```diff
@@ -469,3 +475,223 @@ CREATE TABLE dbo.policies(
+      dttype, -- dtstart dtend type
+      location,
+      week_days, -- by day
+      NULL -- until
```
This is the missing `end_date` of the `CalendarTerm` right?

jmchamica replied:
Yes!

jmchamica replied:
In the previous meeting, a decision was made to implement a dependency between and recurrent event's starting and end dates with the Calendar Term.

alexandru-pena commented on project/src/main/kotlin/org/ionproject/core/writeApi/insertClassSectionEvents/InsertClassSectionEventsController.kt:
```diff
@@ -0,0 +1,106 @@
+    val errMessages = SchemaValidator.validate(json)
+    if (errMessages.isNotEmpty()) {
+      // Invalid JSON received
+      return ResponseEntity.badRequest().body(
```
Wouldn't it be better, to avoid repeating code, adding this as an exception to the CustomExceptions folder?

jmchamica replied:
I'm not sure what you mean. No exception is thrown here.

alexandru-pena commented on project/src/main/kotlin/org/ionproject/core/writeApi/insertClassSectionEvents/InsertClassSectionEventsController.kt:
```diff
@@ -0,0 +1,106 @@
+class InsertClassSectionEventsController(private val repo: InsertClassSectionEventsRepo) {
+
+  @PutMapping(Uri.insertClassSectionEvents, consumes = [Media.APPLICATION_JSON])
+  fun insertClassSectionEvents(@RequestBody json: JsonNode): ResponseEntity<Any> {
```
It may be a personal opinion, but this method is gigantic, maybe try to refactor the sql interactions out of the controller method?

jmchamica replied:
Will do :+1:

jmchamica committed:
Merge branch 'master' into feature/gh-144-insertClassSectionEvents-api

jmchamica committed:
[Write API]: Code refactor. Added Service class.

jmchamica committed:
[ Write API ]: Replace join() with the native joinToString()

jmchamica committed:
[Write API]: Schema and SQL Procs changes.

Accept events with different categories.
Acronyms are mandatory on #academicObjects.

jmchamica committed:
Merge branch 'master' into feature/gh-144-insertClassSectionEvents-api
